File: app/src/main/java/com/majeur/psclient/ChatFragment.java
package com.majeur.psclient;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ScrollView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.majeur.psclient.io.DexIconLoader;
import com.majeur.psclient.io.GlideHelper;
import com.majeur.psclient.model.RoomInfo;
import com.majeur.psclient.service.RoomMessageObserver;
import com.majeur.psclient.service.ShowdownService;
import com.majeur.psclient.util.Utils;
import com.majeur.psclient.util.html.Html;

import java.util.List;

import static com.majeur.psclient.model.Id.toIdSafe;

public class ChatFragment extends Fragment implements MainActivity.Callbacks {

    private InputMethodManager mInputMethodManager;
    private ShowdownService mShowdownService;
    private GlideHelper mSpritesLoader;
    private DexIconLoader mDexIconLoader;

    private TextView mChatTextView;
    private ScrollView mChatScrollView;
    private TextView mTitleView;
    private TextView mUserCountView;
    private ImageButton mTitleButton;
    private EditText mMessageView;
    private ImageButton mSendMessageView;

    private String mObservedRoomId;

    public String getObservedRoomId() {
        return mObservedRoomId;
    }

    public void setObservedRoomId(String observedRoomId) {
        mObservedRoomId = observedRoomId;
        mObserver.setObservedRoomId(observedRoomId);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mSpritesLoader = ((MainActivity) context).getGlideHelper();
        mDexIconLoader = ((MainActivity) context).getDexIconLoader();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mInputMethodManager = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_chat, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mChatTextView = view.findViewById(R.id.text_view_chat);
        mChatTextView.animate().setDuration(200);
        mChatTextView.setMovementMethod(new LinkMovementMethod());
        mChatScrollView = view.findViewById(R.id.scroll_view_chat);
        mTitleView = view.findViewById(R.id.text_view_title);
        mUserCountView = view.findViewById(R.id.text_view_users);
        mTitleButton = view.findViewById(R.id.button_join_room);
        mMessageView = view.findViewById(R.id.edit_text_chat);
        mSendMessageView = view.findViewById(R.id.button_send);

        mTitleButton.setOnClickListener(v -> {
            if (!mShowdownService.isConnected()) return;

            if (mObserver.getRoomJoined())
                mShowdownService.sendRoomCommand(mObservedRoomId, "leave");
            else
                mShowdownService.sendGlobalCommand("cmd", "rooms");
        });

        mUserCountView.setOnClickListener(v -> {
            ArrayAdapter<String> adapter = new ArrayAdapter<>(v.getContext(),
                    android.R.layout.simple_list_item_1,
                    mObserver.getUsers());
            new AlertDialog.Builder(v.getContext())
                    .setTitle("Users")
                    .setAdapter(adapter, (dialog, pos) -> {
                        mShowdownService.sendGlobalCommand("cmd userdetails", toIdSafe(adapter.getItem(pos)));
                        dialog.dismiss();
                    })
                    .setNegativeButton("Close", null)
                    .show();
        });

        mSendMessageView.setOnClickListener(v -> sendMessageIfAny());

        mMessageView.setOnEditorActionListener((textView, actionId, keyEvent) -> {
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                sendMessageIfAny();
                return true;
            }
            return false;
        });

        setDefaultUiState();
    }

    private void setDefaultUiState() {
        mTitleView.setText("—");
        mUserCountView.setText("-\nusers");
        mChatTextView.animate().alpha(0f).withEndAction(() -> {
            mChatTextView.setText("\n\n\n\n\n\n\n\n\n\nTap the join button to join a room");
            mChatTextView.setGravity(Gravity.CENTER_HORIZONTAL);
            mChatTextView.animate().alpha(1f).withEndAction(null).start();
        }).start();
        mMessageView.getText().clear();
        mMessageView.clearFocus();
        mInputMethodManager.hideSoftInputFromWindow(mMessageView.getWindowToken(), 0);
        mMessageView.setEnabled(false);
        mSendMessageView.setEnabled(false);
        mSendMessageView.getDrawable().setAlpha(128);
        mTitleButton.setImageResource(R.drawable.ic_enter);
        mTitleButton.requestFocus();
    }

    private void sendMessageIfAny() {
        String message = mMessageView.getText().toString();
        if (message.length() == 0 || TextUtils.isEmpty(message))
            return;
        mShowdownService.sendRoomMessage(mObservedRoomId, message);
        mMessageView.getText().clear();
        mInputMethodManager.hideSoftInputFromWindow(mMessageView.getWindowToken(), 0);
    }

    @Override
    public void onServiceBound(ShowdownService service) {
        mShowdownService = service;
        service.registerMessageObserver(mObserver);
    }

    @Override
    public void onServiceWillUnbound(ShowdownService service) {
        mShowdownService = null;
        service.unregisterMessageObserver(mObserver);
    }

    void onAvailableRoomsChanged(RoomInfo[] officialRooms, RoomInfo[] chatRooms) {
        JoinRoomDialog.newInstance(officialRooms, chatRooms).show(getFragmentManager(), "");
    }

    private void notifyNewMessageReceived() {
        MainActivity activity = (MainActivity) getActivity();
        if (getId() != activity.getSelectedFragmentId())
            activity.showBadge(getId());
    }

    private final RoomMessageObserver mObserver = new RoomMessageObserver() {

        @Override
        public void onRoomInit() {
            mMessageView.setEnabled(true);
            mMessageView.requestFocus();
            mSendMessageView.setEnabled(true);
            mSendMessageView.getDrawable().setAlpha(255);
            mTitleButton.setImageResource(R.drawable.ic_exit);
            mChatTextView.setGravity(Gravity.START);
            mChatTextView.setText("", TextView.BufferType.EDITABLE);
            postFullScroll();
        }

        @Override
        public void onPrintText(CharSequence text) {
            boolean fullScrolled = Utils.fullScrolled(mChatScrollView);
            int l = mChatTextView.length();
            if (l > 0) mChatTextView.append("\n");
            mChatTextView.append(text);
            notifyNewMessageReceived();
            if (fullScrolled) postFullScroll();
        }

        @Override
        protected void onPrintHtml(String html) {
            final Object mark = new Object();
            int l = mChatTextView.length();
            mChatTextView.append("\u200C");
            mChatTextView.getEditableText().setSpan(mark, l, l + 1, Spanned.SPAN_MARK_MARK);
            Html.fromHtml(html,
                    Html.FROM_HTML_MODE_COMPACT,
                    mSpritesLoader.getHtmlImageGetter(mDexIconLoader, mChatTextView.getWidth()),
                    spanned -> {
                        int at = mChatTextView.getEditableText().getSpanStart(mark);
                        if (at == -1) return; // Check if text has been cleared
                        boolean fullScrolled = Utils.fullScrolled(mChatScrollView);
                        mChatTextView.getEditableText()
                                .insert(at, "\n")
                                .insert(at + 1, spanned);
                        notifyNewMessageReceived();
                        if (fullScrolled) postFullScroll();
                    });
        }

        private void postFullScroll() {
            mChatScrollView.post(() -> mChatScrollView.fullScroll(View.FOCUS_DOWN));
        }

        @Override
        public void onRoomTitleChanged(String title) {
            mTitleView.setText(title);
        }

        @Override
        public void onUpdateUsers(List<String> users) {
            mUserCountView.setText(users.size() + " users");
        }

        @Override
        public void onRoomDeInit() {
            setDefaultUiState();
        }
    };
}

File: app/src/main/java/com/majeur/psclient/ui/ChatFragment.kt
package com.majeur.psclient.ui

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.view.*
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.ArrayAdapter
import android.widget.TextView
import com.majeur.psclient.R
import com.majeur.psclient.databinding.FragmentChatBinding
import com.majeur.psclient.io.AssetLoader
import com.majeur.psclient.io.GlideHelper
import com.majeur.psclient.model.Id
import com.majeur.psclient.model.RoomInfo
import com.majeur.psclient.service.RoomMessageObserver
import com.majeur.psclient.service.ShowdownService
import com.majeur.psclient.util.Callback
import com.majeur.psclient.util.Utils
import com.majeur.psclient.util.html.Html


class ChatFragment : BaseFragment() {

    private lateinit var inputMethodManager: InputMethodManager
    private lateinit var glideHelper: GlideHelper
    private lateinit var assetLoader: AssetLoader

    private var _binding: FragmentChatBinding? = null
    private val binding get() = _binding!!

    private var _observedRoomId: String? = null
    var observedRoomId: String?
        get() = _observedRoomId
        set(observedRoomId) {
            _observedRoomId = observedRoomId
            observer.observedRoomId = observedRoomId
        }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        inputMethodManager = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        glideHelper = mainActivity.glideHelper
        assetLoader = mainActivity.assetLoader
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentChatBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.chatLog.apply {
            movementMethod = LinkMovementMethod()
            animate().duration = 200
        }
        binding.joinButton.setOnClickListener {
            if (service?.isConnected != true) return@setOnClickListener
            if (observer.roomJoined) service?.sendRoomCommand(observedRoomId, "leave")
            else service?.sendGlobalCommand("cmd", "rooms")
        }
        binding.usersCount.setOnClickListener { v: View ->
            val adapter = ArrayAdapter(requireActivity(), android.R.layout.simple_list_item_1, observer.users)
            AlertDialog.Builder(requireActivity())
                    .setTitle("Users")
                    .setAdapter(adapter) { dialog: DialogInterface, pos: Int ->
                        service?.sendGlobalCommand("cmd userdetails", Id.toIdSafe(adapter.getItem(pos)))
                        dialog.dismiss()
                    }
                    .setNegativeButton("Close", null)
                    .show()
        }
        binding.sendButton.setOnClickListener { sendMessageIfAny() }
        binding.messageInput.setOnEditorActionListener { _: TextView?, actionId: Int, _: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                sendMessageIfAny()
                return@setOnEditorActionListener true
            }
            false
        }
        setUiState(roomJoined = false)
    }

    private fun setUiState(roomJoined: Boolean) {
        if (roomJoined) {
            binding.apply {
                messageInput.isEnabled = true
                messageInput.requestFocus()
                sendButton.isEnabled = true
                sendButton.drawable.alpha = 255
                joinButton.setImageResource(R.drawable.ic_exit)
                chatLog.gravity = Gravity.START
                chatLog.setText("", TextView.BufferType.EDITABLE)
            }
        } else {
            binding.apply {
                roomTitle.text = "—"
                usersCount.text = "-\nusers"
                messageInput.text.clear()
                messageInput.clearFocus()
                messageInput.isEnabled = false
                sendButton.isEnabled = false
                sendButton.drawable.alpha = 128
                joinButton.setImageResource(R.drawable.ic_enter)
                joinButton.requestFocus() // Remove focus from message input widget
                chatLog.animate().alpha(0f).withEndAction {
                    chatLog.text = "\n\n\n\n\n\n\n\n\n\nTap the join button to join a room"
                    chatLog.gravity = Gravity.CENTER_HORIZONTAL
                    chatLog.animate().alpha(1f).withEndAction(null).start()
                }.start()
            }
            inputMethodManager.hideSoftInputFromWindow(binding.messageInput.windowToken, 0)
        }
    }

    private fun sendMessageIfAny() {
        val message = binding.messageInput.text.toString()
        if (message.isEmpty()) return
        service?.sendRoomMessage(observedRoomId, message)
        binding.messageInput.text.clear()
    }

    override fun onServiceBound(service: ShowdownService) {
        super.onServiceBound(service)
        service.registerMessageObserver(observer)
    }

    override fun onServiceWillUnbound(service: ShowdownService) {
        super.onServiceWillUnbound(service)
        service.unregisterMessageObserver(observer)
    }

    fun onAvailableRoomsChanged(officialRooms: List<RoomInfo>, chatRooms: List<RoomInfo>) {
        JoinRoomDialog.newInstance(officialRooms.toTypedArray(), chatRooms.toTypedArray())
                .show(requireFragmentManager(), "")
    }

    private fun notifyNewMessageReceived() {
        mainActivity.showBadge(id)
    }

    private fun postFullScroll() {
        binding.chatLogContainer.post { binding.chatLogContainer.fullScroll(View.FOCUS_DOWN) }
    }

    private val observer: RoomMessageObserver = object : RoomMessageObserver() {

        public override fun onRoomInit() {
            setUiState(roomJoined = true)
        }

        public override fun onRoomDeInit() {
            setUiState(roomJoined = false)
        }

        public override fun onPrintText(text: CharSequence) {
            val fullScrolled = Utils.fullScrolled(binding.chatLogContainer)
            if (binding.chatLog.length() > 0) binding.chatLog.append("\n")
            binding.chatLog.append(text)
            notifyNewMessageReceived()
            if (fullScrolled) postFullScroll()
        }

        override fun onPrintHtml(html: String) {
            val mark = Any()
            val l = binding.chatLog.length()
            binding.chatLog.append("\u200C")
            binding.chatLog.editableText.setSpan(mark, l, l + 1, Spanned.SPAN_MARK_MARK)
            Html.fromHtml(html,
                    Html.FROM_HTML_MODE_COMPACT,
                    glideHelper.getHtmlImageGetter(assetLoader, binding.chatLog.width),
                    Callback { spanned: Spanned? ->
                        val at = binding.chatLog.editableText.getSpanStart(mark)
                        if (at == -1) return@Callback // Check if text has been cleared
                        val fullScrolled = Utils.fullScrolled(binding.chatLogContainer)
                        binding.chatLog.editableText
                                .insert(at, "\n")
                                .insert(at + 1, spanned)
                        notifyNewMessageReceived()
                        if (fullScrolled) postFullScroll()
                    })
        }

        public override fun onRoomTitleChanged(title: String) {
            binding.roomTitle.text = title
        }

        public override fun onUpdateUsers(users: List<String>) {
            binding.usersCount.text = "${users.size} users"
        }
    }
}